src/core/OIDNDenoiser.js

```diff
@@ -0,0 +1,298 @@
+
+	}
+
+	set cleanAux( v ) {
+
+		this.denoiser.dirtyAux = ! v;
+
+	}
+
+	get hdr() {
+
+		return this.denoiser.hdr;
+
+	}
+
+	set hdr( v ) {
+
+		this.denoiser.hdr = v;
+		if ( v && this.quality === 'fast' ) this.quality = 'balanced';
+
+	}
+
+	get weightsUrl() {
+
+		return this.denoiser.weightsUrl;
+
+	}
+
+	set weightsUrl( url ) {
+
+		this.denoiser.weightsUrl = url;
+
+	}
+
+	//* Debugging
+	get denoiserDebugging() {
+
+		return this.denoiser.debugging;
+
+	}
+
+	set denoiserDebugging( v ) {
+
+		this.denoiser.debugging = v;
+
+	}
+
+	constructor( renderer ) {
+
+		this.renderer = renderer;
+		this.denoiser = new Denoiser( 'webgl', renderer.domElement );
+		this.denoiser.inputMode = 'webgl';
+		this.denoiser.outputMode = 'webgl';
+		this.denoiser.weightsUrl = 'https://cdn.jsdelivr.net/npm/denoiser/tzas';
+		this.denoiser.onProgress( progress => this.handleProgress( progress ) );
+		//this.denoiser.hdr = true;
+
+		this.isDenoising = false;
+		this.fadeTime = 500;
+		this.denoiserFinished = 0;
+		this.cleanAux = true;
+		this._useAux = true;
+		this.externalAux = false;
+		this.auxTextures = { albedo: null, normal: null };
+		this.progressListeners = new Set();
+
+		this.linearToSRGBMaterial = new MeshBasicMaterial( {
+			map: null,
+			transparent: true,
+			blending: NoBlending,
+		} );
+
+		// Material to blend between pathtracer and denoiser
+		this.blendToCanvasMaterial = new ClampedInterpolationMaterial( {
+			map: null,
+			transparent: true,
+			premultipliedAlpha: renderer.getContextAttributes().premultipliedAlpha,
+		} );
+
+
+		this.quad = new FullScreenQuad( this.linearToSRGBMaterial );
+		this.createConversionRenderTarget( renderer.domElement.width, renderer.domElement.height );
+
+	}
+
+	setAuxTextures( albedoTexture, normalTexture ) {
+
+		this.externalAux = true;
+		this.auxTextures.albedo = albedoTexture;
+		this.auxTextures.normal = normalTexture;
+
+	}
+
+	async denoise( rawPathtracedTexture, albedoTexture, normalTexture ) {
+
+		this.isDenoising = true;
+
+		// Adjust the height /width if changed from before
+		const { width, height } = rawPathtracedTexture.image;
+		if ( this.denoiser.height !== height || this.denoiser.width !== width ) {
+
+			this.denoiser.width = width;
+			this.denoiser.height = height;
+			this.createConversionRenderTarget( width, height );
+
+		}
+
+		// set this on the object so we can get it later to split it
+		this.pathtracedTexture = this.getLinearToSRGBTexture( rawPathtracedTexture, this.conversionRenderTarget );
+
+		// Extract the raw webGLTextures
+		const colorWebGLTexture = this.getWebGLTexture( this.pathtracedTexture );
+		const albedoWebGLTexture = this.getWebGLTexture( albedoTexture );
+		const normalWebGLTexture = this.getWebGLTexture( normalTexture );
+
+		//* run the denoiser ----------------------------
+		this.renderer.resetState();
+
+		// Run the denoiser
+		const denoisedWebGLTexture = await this.denoiser.execute( colorWebGLTexture, albedoWebGLTexture, normalWebGLTexture );
```

Comment:
This looks like a memory leak - does the model always return a brand new texture? Do we ever dispose of the webgl memory?

Reply:
The Albedo and Normal ones are just references to the original threejs textures so get disposed when those textures would. Not sure how that works in three, either when the renderTarget renders etc.

The `WebGLTexture` out of the denoiser is bound to the `outputTensor` of the denoiser flow. When the next model is executed and a new output is generated, the old output is disposed, destroying the bound texture.

Reply:
The albedo and normal textures I understand - the returned handle from `denoiser.execute` is what I was wondering about. It sound like the life cycle of this handle is internally managed, then.

Reply:
Yeah, I'll try to explain better.
To get the WebGL Texture we run something like:
```ts
const outputWebGLTexture = outputTensor.dataToGPU();
```

This `WebGLTexture` is what is bound to the tensor in the backend and what we inject into threejs. 

when the denoiser is executed again I destroy the old output tensor which destroys the WebGLTexture it is bound to.
So no worries here.